C++ warnings from compiling pileup code

With

% g++ --version
g++ (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

I get the following compilation warnings from `sherpa/astro/utils/src/pileup.cc`

```
g++: sherpa/astro/utils/src/pileup.cc
sherpa/astro/utils/src/pileup.cc: In function ‘int apply_pileup(unsigned int, const double*, double*, double*, double*, double, unsigned int, unsigned int*, const double*, const double*, const double*, double, double, double, double, double, double, sherpa::usrfuncproto, sherpa::PyWrapper*)’:
sherpa/astro/utils/src/pileup.cc:509:35: warning: ‘k.pileup_kernel_t::arf_s_tmp’ may be used uninitialized in this function [-Wmaybe-uninitialized]
    if (k->arf_s_tmp != NULL) free (k->arf_s_tmp);
                              ~~~~~^~~~~~~~~~~~~~
sherpa/astro/utils/src/pileup.cc:606:19: note: ‘k.pileup_kernel_t::arf_s_tmp’ was declared here
   pileup_kernel_t k;
                   ^
sherpa/astro/utils/src/_pileup.cc: In function ‘PyObject* _apply_pileup(PyObject*, PyObject*)’:
sherpa/astro/utils/src/_pileup.cc:126:8: warning: ‘template<class> class std::auto_ptr’ is deprecated [-Wdeprecated-declarations]
   std::auto_ptr< sherpa::PyWrapper >
        ^~~~~~~~
In file included from /usr/include/c++/7/memory:80:0,
                 from sherpa/astro/utils/src/_pileup.cc:22:
/usr/include/c++/7/bits/unique_ptr.h:51:28: note: declared here
   template<typename> class auto_ptr;
                            ^~~~~~~~
```

Changing `std::auto_ptr` to `std::unique_ptr`` stops the warning message. You need to have XSPEC (and presumably a FITS backend) to actually check this fix - my current run looks to have failed with 

    @requires_fits
    @requires_xspec
    def test_pileup(self):
        self.run_thread('pileup')
    
        fr = ui.get_fit_results()
        covarerr = sqrt(fr.extra_output['covar'].diagonal())
>       assert covarerr[0] == approx(684.056 , rel=1e-4)
E       AssertionError: assert 694.3746848979197 == 684.056 ± 6.8e-02
E        +  where 684.056 ± 6.8e-02 = approx(684.056, rel=0.0001)

but that may not be due to to the pointer change but some other thing (strange things are afoot in this build!).

However, looking at the code it's not clear why we need this wrapper object, since all it seems to be used for is passing a function pointer between functions. So maybe we can just remove it completely?

--- a/sherpa/astro/utils/src/_pileup.cc
+++ b/sherpa/astro/utils/src/_pileup.cc
@@ -1,5 +1,5 @@
 // 
-//  Copyright (C) 2009  Smithsonian Astrophysical Observatory
+//  Copyright (C) 2009, 2021  Smithsonian Astrophysical Observatory
 //
 //
 //  This program is free software; you can redistribute it and/or modify
@@ -123,8 +123,13 @@ static PyObject* _apply_pileup( PyObject* self, PyObject* args )
     return NULL;
   }
 
+#if (__cplusplus < 201103L)
   std::auto_ptr< sherpa::PyWrapper >
     wrapper( new sherpa::PyWrapper( py_function ) );
+#else
+  std::unique_ptr< sherpa::PyWrapper >
+    wrapper( new sherpa::PyWrapper( py_function ) );
+#endif
 
   DoubleArray results;
   if ( EXIT_SUCCESS != results.create( arf_source.get_ndim(),

